.github/workflows/version-or-release.yml

```diff
@@ -0,0 +1,143 @@
+          script: require('./package.json').version
+
+      - name: Check if ${{ steps.latest-version.outputs.result }} branch is already created
+        id: release-branch-check
+        run: |
+          branch=release-temp-${{ steps.latest-version.outputs.result }}
+
+          if [ "$(git ls-remote https://github.com/${{ github.repository }}.git $branch)" == "" ]; then
+            echo "$branch doesn't exist on the remote"
+            echo "::set-output name=branch::$branch"
+          else
+            echo "$branch already exists on the remote"
+          fi;
+
+  staging-deploy-trigger:
+    name: Push release to staging
+    runs-on: ubuntu-latest
+    needs: create-temp-release-branch
+    environment: Staging Trigger
+
+    steps:
+      - name: Checkout Repo
+        uses: actions/checkout@v2
+        with:
+          ref: ${{ needs.create-temp-release-branch.outputs.release-branch }}
+
+      - name: Setup git user
+        run: |
+          git config user.name "github-actions[bot]"
+          git config user.email "github-actions[bot]@users.noreply.github.com"
+
+      - name: Create release commit
+        run: |
+          git commit -m 'Release ${{ needs.create-temp-release-branch.outputs.version }} to staging' --allow-empty
+
+      - name: Force push the release to staging
+        run: |
+          git push origin HEAD:staging --force
+
+  production-deploy-trigger:
+    name: Push release to main
+    runs-on: ubuntu-latest
+    needs: create-temp-release-branch
+    environment: Production Trigger
+
+    steps:
+      - name: Checkout Repo
+        uses: actions/checkout@v2
+        with:
+          ref: ${{ needs.create-temp-release-branch.outputs.release-branch }}
+
+      - name: Setup git user
+        run: |
+          git config user.name "github-actions[bot]"
+          git config user.email "github-actions[bot]@users.noreply.github.com"
+
+      - name: Create release commit
+        run: |
+          git commit -m 'Release ${{ needs.create-temp-release-branch.outputs.version }} to main' --allow-empty
+
+      - name: Force push the release to main
+        run: |
+          git push origin HEAD:main --force
+
+  create-github-release:
+    name: Create GitHub Release
+    runs-on: ubuntu-latest
+    needs: [create-temp-release-branch, production-deploy-trigger]
+
+    steps:
+      - name: Checkout Repo
+        uses: actions/checkout@v2
+        with:
+          ref: ${{ needs.create-temp-release-branch.outputs.release-branch }}
+
+      - name: Publish a tag
+        run: |
+          git tag v${{ needs.create-temp-release-branch.outputs.version }}
+          git push origin v${{ needs.create-temp-release-branch.outputs.version }}
+
+      - name: Publish a GitHub release
+        uses: actions/github-script@v4
+        with:
+          script: |
+            const { publishRelease } = require(`./scripts/publish-github-release.js`)
+
+            await publishRelease(github, { repo: context.repo })
```

Comment:
Refactor item: can we put this in a script file?

Reply:
Right now the `publish-github-release.js` is more reusable, which won't matter for this one - but in the previous job I've used this pattern to reuse a single script by different workflows, where each workflow was able to parametrize that script how they wanted. If we ever add any such shared script then keeping this one like this makes it more consistent with other, potential, scripts.

In addition to that, providing the `github` (aka `octokit`) allows us to skip installing deps - because we just grab what has been provided to this action and pass and inject it to the script. So it's a small resource optimization